File: haskell/tests/Test/Free.hs
module Test.Free (test) where

import Free.App (maybeDouble)
import Test.Free.Interpreters (runWithRiggedCoin)
import Test.Hspec (SpecWith, context, describe, it, shouldBe)

-- With the test interpreter testing the core logic is a breeze!
test :: SpecWith ()
test = do
  let program = maybeDouble 10
  describe "maybeDouble" $ do
    context "when executed" $ do
      it "should output a message" $ do
        let (_, msg1) = program `runWithRiggedCoin` True
        let (_, msg2) = program `runWithRiggedCoin` False
        msg1 `shouldBe` "Flipping a coin!"
        msg2 `shouldBe` "Flipping a coin!"

    context "when the coin flip result is heads" $ do
      it "should double the number" $ do
        let (res, _) = program `runWithRiggedCoin` True
        res `shouldBe` 20

    context "when the coin flip result is tails" $ do
      it "should not double the number" $ do
        let (res, _) = program `runWithRiggedCoin` False
        res `shouldBe` 10

File: scala/src/test/scala/free/Free.scala
package free

import org.scalatest.wordspec.AnyWordSpec
import org.scalatest.matchers.should.Matchers
import free.core.maybeDouble
import free.interpreters.Test.*

// With the test interpreter testing the core logic is a breeze!
class Test extends AnyWordSpec, Matchers:
  val program = maybeDouble(10)
  "maybeDouble" when {
    "executed" should {
      "always print a message regardless of the coin flip result" in {
        val (msg1, _) = program.runWithRiggedCoin(true)
        val (msg2, _) = program.runWithRiggedCoin(false)
        msg1 shouldBe "Flipping a coin!"
        msg2 shouldBe "Flipping a coin!"
      }
    }
    "the coin flip result is heads" should {
      "double the number" in {
        val (_, res) = program.runWithRiggedCoin(true)
        res shouldBe 20
      }
    }
    "the coin flip result is tails" should {
      "not double the number" in {
        val (_, res) = program.runWithRiggedCoin(false)
        res shouldBe 10
      }
    }
  }
